Added a TS/RA2 sprite sequence importer for infantry and buildings

Mailaender opened the pull request:
These are traditionally time consuming to do manually.

Mailaender committed:
Add a sprite sequence importer for infantry and buildings.

abcdefg30 commented on OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs:
```diff
@@ -0,0 +1,197 @@
+		{
+			Console.WriteLine(section.Name + ":");
+
+			Console.WriteLine("\tDefaults:");
```
Won't this create an "empty" `Defaults:` "section" in the yaml, if the imported structure doesn't use tileset extensions?

But anyway, I guess even if, this shouldn't hurt.

Mailaender replied:
No tileset codes/extensions is quite the exception for TS/RA2 mods. These auto-generated rules need a cleanup afterwards anyway.

abcdefg30 commented:
Hm, seems like this is generation `Facings: 0` values, which is not correct (e.g. import the sequences of the rocketeer and look at its idle definitions) and on other sequences it used `Facings: 6` where it should be 8 (probably not your fault, as the value is really 6, but the assets browser shows me 8 facings...).

GJ otherwise.

Mailaender commented:
It indeed has some quirks. For example the `stand` sequence always needs to be corrected to:

```diff
	stand:
-		Start: 0
-		Length: 1
-		Facings: 1
	stand:
+		Start: 0
+		Length: 6
+		Facings: 8
```

Those definitions seem to have different meanings for each type although the enumeration looks identical.

GraionDilach commented on OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs:
```diff
@@ -0,0 +1,197 @@
+			{
+				Console.WriteLine("\t\tStart: " + splitting[0]);
+				Console.WriteLine("\t\tLength: " + splitting[1]);
+				Console.WriteLine("\t\tFacings: " + splitting[2]);
```
`splitting.Length == 4 ? Console.WriteLine("\t\tFacings: 1") : Console.WriteLine("\t\tFacings: 8")` please.

If I get it right, `splitting[2] - splitting[1]` would be OpenRA `Stride:` but I never seen infantry which either used this logic either used the `Stride` tag in OpenRA.

GraionDilach commented:
That 0 1 1 stand looks legit though, TS/RA2 couldn't use more frames for idling at all, jumpjet infantry are a giant hack itself, your 0 6 8 fix sounds wrong to me instead tbh. `Length` shouldn't be increased.

Mailaender commented:
Thanks, added your suggested fixes and some more infantry sequences such as die, panic and cheer.

abcdefg30 commented:
:+1: Looks good to me now.

obrakmann commented on OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs:
```diff
@@ -0,0 +1,253 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
```
I can't comment on the functionality, but you should use the updated copyright header here. It's also 2016 :)

pchote commented:
Any progress here?

GraionDilach commented:
Technically this doesn't need a progress (excluding the header). I think this is done but haven't found time to check the output to give my thumb.

pchote commented:
The header is what i'm referring to.

Mailaender commented:
Updated with the 2016 copyright header.

GraionDilach commented:
You missed `LegacySequenceImporter`.

Mailaender commented:
Fixed.

pchote commented:
You have access to the footprint information, so really should generate the appropriate artwork offsets for buildings (or at least something that is approximately correct).  Ignoring them here and then doing them all by hand (https://github.com/OpenRA/ra2/issues/159) is a giant waste of effort and potential.

Mailaender commented:
Really, I thought that information just wasn't available and all the TS offsets were corrected manually?

pchote commented:
It's not available to the OpenRA sprite sequence parsers, because it depends on the building footprints in the rules.  You have access to the building foundation here.

GraionDilach commented:
For reference:

In TS/RA2 the center of the SHP was the building's top-right (in isometric perspective, the top corner) cells top pixel. OpenRA assumes the center of the SHP is the building's middle.

Besides the footprints, it is based on the actual cell's size, and the offset math is also quite complex at non-square footprints (y offset will always be cellheight*y/2 but the x I'm not certain about, at square footprints, it's 0, but notsure at the rest, might be cellwidth*(x-y)/2 but I'm just guessing).

Including this here might be more complicated than the rest which are universal and shared between TS and RA2 which is why I felt that's being out-of-scope here.

GraionDilach commented:
Attaching a picture from SHP Builder with TS cellsize to point this out further - although it's known SHP Builder mispositions this grid some pixels away, probably due to rounding errors.

![shpbuildertsgrid](https://cloud.githubusercontent.com/assets/1136302/13898900/b1ab53e4-ede0-11e5-9a1e-51ca284c36a3.png)

reaperrr commented:
Adding support for automatic building offset generation would be nice.
Needs a rebase in any case.

Mailaender committed:
Guess the correct sprite offset from building foundations.

Mailaender commented:
Rebased and added the offset guessing, which is indeed an excellent idea and your suggestions work rather good in practice https://github.com/OpenRA/ra2/issues/159 as most houses have a square footprint.

pchote commented:
The offsets are imported incorrectly for all the rectangular footprints that I checked in TS.

Mailaender commented:
I don't see any pattern to be honest:

Dimensions: 4,3

```patch
gaweap:
	Defaults:
-		Offset: 0,-36
+		Offset: -12, -42
```

Dimensions: 3,2

```patch
nahand:
	Defaults:
-		Offset: 0,-24
+		Offset: -6, -30
```
	
Dimensions: 2,3
	
```patch
naapwr:
	Defaults:
-		Offset: -24,-36
+		Offset: 12, -30
```

pchote commented:
Your best test cases will be the TS gates, which are currently imported with an offset that is wrong by multiple cells.

Mailaender commented:
Dimensions: 3,1

```patch
gagate_a:
	Defaults:
-		Offset: 48,-12
+		Offset: -24, -24
```

Dimensions: 1,3

```patch
gagate_b:
	Defaults:
-		Offset: -48,-36
+		Offset: 24, -24
```

Mailaender commented:
Removed https://github.com/OpenRA/OpenRA/pull/10715/commits/718107bbc50003398141957655ed6bb1f4e45641, because while this is a good idea, it is a large enough feature for its own pull request and this one already in development for 4 months.

pchote commented:
From above:
> You have access to the footprint information, so really should generate the appropriate artwork offsets > for buildings (or at least something that is approximately correct). Ignoring them here and then doing
> them all by hand (OpenRA/ra2#159) is a giant waste of effort and potential for this tool.

I don't think it is fair to ask us to merge an unfinished tool upstream just because you don't want to work through the equations to finish it.  That *forces* the burden of effort onto someone else, compared with *asking* for help and receiving it on their timescale.

Mailaender commented:
In fact I removed the commit to not submit unfinished code, but only proven one.

pchote commented:
Do you have a branch or commit sha1 for the version with the incorrect offset calculation?

pchote commented:
`FootprintUtils.CenterOffset` gives you the offset between the center of the top-left cell and the center of the actor.  Just add 512 to Y and then multiply by the tile pixel size / 1024.

Mailaender commented:
Updated with an improved version of the building offset converter thanks to your help. I just changed the sign and divided the X offset by 2. Updated https://github.com/OpenRA/ra2/pull/224 where the map "Powder Keg" is a good test case, because it features a lot of buildings that are aligned to the street tiles.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs:
```diff
@@ -0,0 +1,275 @@
+			Console.WriteLine("\tDefaults:");
+
+			var foundation = section.GetValue("Foundation", string.Empty);
+			if (!string.IsNullOrEmpty(foundation))
```
You can simplify this much more:
```csharp
			if (!string.IsNullOrEmpty(foundation))
			{
				var size = foundation.Split('x');
				if (size.Length == 2)
				{
					var x = int.Parse(size[0]);
					var y = int.Parse(size[1]);
	
					var xOffset = (x - y) * grid.TileSize.Width / 4;
					var yOffset = (x + y) * grid.TileSize.Height / 4;
					Console.WriteLine("\t\tOffset: {0},{1}", -xOffset, -yOffset);
				}
			}
```

Mailaender replied:
Added. This won't hide as you commented on a line that isn't affected by your suggestions.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs:
```diff
@@ -0,0 +1,275 @@
+				Console.WriteLine("\t\tStart: " + splitting[0]);
+				Console.WriteLine("\t\tLength: " + splitting[1]);
+
+				// TODO splitting[2] - splitting[1] may be Stride:
```
Do you have an example of where splitting[2] is used?

GraionDilach replied:
https://github.com/OpenRA/OpenRA/pull/10715#discussion_r53581402

GraionDilach replied:
Oh, apparently there are cases in RA2 where this is used (crawl's first frames are reused as prone in RA2). https://github.com/Phrohdoh/oramod-ra2/blob/master/extra-files/art.ini#L9637

pchote replied:
Ok, then we should resolve this easy TODO before merging.

Mailaender replied:
I have no idea if I understand what this strides is, but I added it.

pchote commented:
This now spits out `Stride: 0` for everything that doesn't need it, and all the non-zero values are negative.

pchote commented:
For definitions with 3 entries splitting[2] appears to be the stride, and should only be written if it is different to splitting[1].

Mailaender commented:
Updated the stride calculation with your recent findings.

pchote commented:
I think i've done enough nit-picking now.  The generated sequences look sensible, so :+1:.

pchote commented:
This feature isn't shipped, so omitting changelog (and removing the entry for the rules importer).
